core/shared/src/main/scala/cats/parse/Parser.scala

```diff
@@ -2144,29 +2157,31 @@ object Parser {
           state.offset += message.length
           ()
         } else {
-          state.error = Chain.one(Expectation.OneOfStr(offset, message :: Nil))
+          state.error = Eval.later(Chain.one(Expectation.OneOfStr(offset, message :: Nil)))
           ()
         }
       }
     }
 
     case class Fail[A]() extends Parser[A] {
       override def parseMut(state: State): A = {
-        state.error = Chain.one(Expectation.Fail(state.offset));
+        val offset = state.offset
+        state.error = Eval.later(Chain.one(Expectation.Fail(offset)))
         null.asInstanceOf[A]
       }
     }
 
     case class FailWith[A](message: String) extends Parser[A] {
       override def parseMut(state: State): A = {
-        state.error = Chain.one(Expectation.FailWith(state.offset, message));
+        val offset = state.offset
+        state.error = Eval.later(Chain.one(Expectation.FailWith(offset, message)))
         null.asInstanceOf[A]
       }
     }
 
     final def oneOf[A](all: Array[Parser0[A]], state: State): A = {
       val offset = state.offset
-      var errs: Chain[Expectation] = Chain.nil
+      var errs: Eval[Chain[Expectation]] = Eval.later(Chain.nil)
```

Comment:
can we make this `Eval.now(Chain.nil)` I think that will be less allocation and actually every `oneOf` hits this path.